### PR TITLE
Eliminate all uses of the page table from the GC (for no-naked-pointers mode)

### PR DESCRIPTION
**Pull request by mshinwell**
The first part of this patch adds GC block headers in the text segments immediately prior to function entry points.  This enables the GC to safely stop scanning when code pointers are reached in no-naked-pointers mode, rather than requiring a page table test, or indeed any special handling for closures.  (Another option would be to alter the closure representation to make it possible to identify which fields of a closure contain code pointers, but you then still need a special case in the GC.)  I think I've convinced myself this is safe.  In addition to being a useful optimization now, this work is a prerequisite for the multicore branch.

The second part of this patch eliminates all other references to the page table from the GC in no-naked-pointers mode.  Damien, I'd appreciate it if you could check this part.

### CODE REVIEW

**Comment by alainfrisch** [comment]
You should probably base this work on trunk, because of the pervasive changes in emit.mlp for i386/amd64 (and the merger between emit.ml/emit_nt.mlp on those architectures).

**Comment by mshinwell** [comment]
alainfrisch: Agreed, but multicore is based on 4.02 at the moment, so we'll need both patches.

Incidentally, for multicore, it looks like we're going to have to add a new tag (Code_tag) so that code pointers can be distinguished in the absence of the page table.

File: byterun/major_gc.c [hunk 1]
```diff
@@ -169,11 +156,8 @@ static void mark_slice (intnat work)
           child = Field (v, i);
 #ifdef NATIVE_CODE_AND_NO_NAKED_POINTERS
           if (Is_block (child)
-                && Wosize_val (child) > 0  /* Atoms never need to be marked. */
-                /* Closure blocks contain code pointers at offsets that cannot
-                   be reliably determined, so we always use the page table when
-                   marking such values. */
-                && (!marking_closure || Is_in_heap (child))) {
+                /* Atoms never need to be marked. */
+                && Wosize_val (child) > 0) {
```
Comment (damiendoligez) [inline byterun/major_gc.c]: Why not make atom headers black and get rid of this test? Then you can use `Nnp_or_is_in_heap` and remove the `#ifdef` entirely.

**Comment by damiendoligez** [comment]
I have reviewd the GC changes and I believe they are correct.

**Comment by alainfrisch** [comment]
> Agreed, but multicore is based on 4.02 at the moment, so we'll need both patches.

I don't think there are plans to create another release on the 4.02 branch.   So I don't see the point of maintaining two patches.  Unless people plan to push hard to multicore part of a (bug-fix -- haha) minor release...

I can see the benefit of providing multicore and your current proposal for 4.02 so that people can test them more easily, but do you agree that there is no point merging the current 4.02-based version?

**Comment by damiendoligez** [comment]
@mshinwell 

> Incidentally, for multicore, it looks like we're going to have to add a new tag (Code_tag) so that code pointers can be distinguished in the absence of the page table.

This means decreasing the number of normal tags by 4, so the maximum number of constant constructors in a type will go from 246 to 242. This is probably not a big deal, but we'll have to check if it breaks some user programs.

**Comment by mshinwell** [comment]
alainfrisch: I wasn't necessarily proposing merging this patch into 4.02; I can make another one for trunk.  However, the decision has been taken that the multicore branch isn't going to be rebased to trunk for some significant period of time, so it made sense to develop against 4.02.

**Comment by mshinwell** [comment]
Xavier: are you OK in principle with this patch (without the change to add Code_tag), for trunk?
It only affects things when the compiler is configured to disallow naked pointers.

**Comment by xavierleroy** [comment]
Just to spell out the obvious: if you're not adding Code_tag headers before code, you need a special case for scanning closures, right?  Because I don't see this in the commits.

I haven't said "no" yet to Code_tag headers.  It's just that we need more (performance) information to choose between that and alternative closure scanning strategies.

**Comment by mshinwell** [comment]
When I said "without the change to add Code_tag", I meant: doing what the patch does at the moment, namely writing headers into the text section, but with Abstract_tag.

**Comment by mshinwell** [comment]
(And I don't think any special case is needed in the GC with the patch in its current state.)

**Comment by xavierleroy** [comment]
Ah, OK, I misunderstood.  Still, I'm curious to see how much performance impact this has, because of the "bubbles" introduced in the code cache.

**Comment by gasche** [comment]
(pinging in case Mark would be interested in having some benchs ready in time for the meeting)

**Comment by lefessan** [comment]
Yes, I would be interested too in cache-related measurements, to compare with specific code for scanning closures. These headers in the code might trigger invalidations of lines in the instruction cache when these lines are loaded in the data cache for header-checks, but is it frequent enough to have an impact on performances ?

**Comment by mshinwell** [comment]
I would be interested in having some benchmarks available, but I doubt I will have time to do so ready for the meeting.  However this could still be approved pending such benchmarking.

**Comment by gasche** [comment]
We discussed this change at the meeting. There was agreement that keeping the pagetable check for function pointers would be a safer approach that preserves most of the general benefits of the patch. My notes are not completely clear on what we decided to do for 4.03 (it was proposed to be conservative and keep it opt-in for 4.03, but no strong consensus; we may need a decision before freeze time in about a month), but it was noted that better tools to help users transition to the no-naked-pointers work would be needed.

**Comment by mshinwell** [comment]
I said I would try to write some kind of checking tool with a view to potentially making no-naked-pointers the default in a future release (not 4.03).

**Comment by damiendoligez** [comment]
I believe the tool already exists (almost):
Configure with `-no-naked-pointers -with-debug-runtime`, then

```
setenv OCAMLPARAM=runtime-variant=d
setenv OCAMLRUNPARAM=v=0
```

Then compile OCaml and your program. Any non-black naked pointer will trigger an assertion.
This doesn't work well today because the GC is still verbose even with `v=0`, but the GC latency patch will change that.

**Comment by damiendoligez** [comment]
I believe some architectures do not support the headers-in-code strategy, so this will need more work. @xavierleroy what is the latest word on this?

**Comment by mshinwell** [comment]
@damiendoligez : Which architectures, do you know?

**Comment by damiendoligez** [comment]
IIRC it was some version of IBM's POWER. I don't remember the OS.

**Comment by mshinwell** [comment]
Oh, I thought the trouble @xavierleroy had for ppc64 could have been _solved_ by a similar approach to this (for a slightly different problem).  I may be misremembering though.

**Comment by mshinwell** [comment]
@xavierleroy Ping on this one please.  We still need to do performance testing for the "code tag" approach, but do you know of any architecture where doing that is going to be impossible for some reason (as per @damiendoligez 's question above)?

**Comment by mshinwell** [comment]
(I am in the process of rebasing this to trunk.)

**Comment by DemiMarie** [comment]
If we are going to have three spare tags, I would like to use them in some way.  Specifically, I would like to add `Mixed_tag`, which would correspond to blocks that consist of **both** pointers and non-pointers.  This could allow unboxing values in every record type with a monomorphic field, much like `{-# UNPACK #-}` in GHC.

**Comment by silene** [comment]
I don't know if it is the best place to mention it, but it would be great if the `Nnp_or_is_in_value_area` macro was used in more places than just the GC. For Coq, we are seriously considering applying (kind of) idiotic pull requests (coq/coq#434) due to the large impact of `Is_in_value_area` on performances.

**Comment by mshinwell** [comment]
I've rebased this patch on top of GPR#756, which cleans things up somewhat.

I believe I have a solution for POWER ELF64v1 ABI systems which involves using the optional third word in the official procedure descriptor (`.opd`) array as a GC header.  It is being tested at the moment.

I have set up benchmarks on x86-64 to see if there are any noticeable performance implications of this.  Once that has been done I will also remove some more `Is_in_value_area` tests in favour of `Nnp_or_is_in_value_area`.

**Comment by xavierleroy** [comment]
two and a half years later, my understanding is that the "headers in code" approach was found to degrade performance noticeably, and that we have decided to move towards a self-describing function closure representation instead.

Based on this understanding, I'm closing this PR.  Yell at me and reopen if I'm wrong.

**Pull request closed without merging**